lnrpc/routerrpc/router_server.go

```diff
@@ -771,11 +771,66 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
 				return err
 			}
 
+		case <-s.quit:
+			return errServerShuttingDown
+			// If the subscribe client terminates, exit with an error.
+		case <-subscription.Quit():
+			return errors.New("htlc event subscription terminated")
+		case <-stream.Context().Done():
+			log.Debugf("Payment stream %v canceled", identifier)
+			return stream.Context().Err()
+		}
+	}
+}
+
+// SubscribePayments returns a stream of payment state updates.
+func (s *Server) SubscribePayments(request *SubscribePaymentsRequest,
+	stream Router_SubscribePaymentsServer) error {
+
+	log.Debug("SubscribePayments called")
+
+	router := s.cfg.RouterBackend
+
+	// Subscribe to payments.
+	subscription, err := router.Tower.SubscribeAllPayments()
+	if err != nil {
+		return err
+	}
+	defer subscription.Close()
+
+	// Stream updates back to the client.
+	for {
+		select {
+		case item, ok := <-subscription.Updates:
```

Comment:
Is there potential for code reuse here from the single payment subscription?

Reply:
It's the interfaces `Router_TrackPaymentV2Server` and `Router_TrackPaymentsServer`
So `stream.Send(rpcPayment)` looks the same, but it's actually two different things.
That got me to copy the thing.

Another option would be to merge TrackPaymentV2 and TrackPayments completely, by allowing the passed hash to be null. But I don't think that's a good idea because they do different things. TrackPayments tracks any currently in-flight payment, while TrackPaymentV2 gets you the status of a specific payment, regardlesss whether it's in-flight or not.

Reply:
No, I wouldn't overload the existing rpc method. But you can extract common code by passing references to the `Send` function and `Done` channel perhaps?

Reply:
Done some code reuse there